### PR TITLE
Feature python cypher

### PR DESCRIPTION
**Pull request by lloydzhou**

### CODE REVIEW

**Comment by knightast** [comment]
CI Failed？

**Comment by hjk41** [comment]
seems to be a lint problem in python_api.cpp
```
 > [stage-0 5/5] RUN --mount=type=secret,id=CODECOV_TOKEN     export CODECOV_TOKEN=$(cat /run/secrets/CODECOV_TOKEN) &&     bash /root/tugraph-db/ci/github_ci.sh:
#10 27.39 src/python/python_api.cpp:62:  Lines should be <= 100 characters long  [whitespace/line_length] [2]
#10 27.39 src/python/python_api.cpp:62:  You don't need a ; after a }  [readability/braces] [4]
#10 27.39 src/python/python_api.cpp:62:  Constructors callable with one argument should be marked explicit.  [runtime/explicit] [5]
#10 27.39 src/python/python_api.cpp:65:  Lines should be <= 100 characters long  [whitespace/line_length] [2]
#10 27.39 src/python/python_api.cpp:65:  You don't need a ; after a }  [readability/braces] [4]
#10 27.39 src/python/python_api.cpp:67:  You don't need a ; after a }  [readability/braces] [4]
#10 27.39 src/python/python_api.cpp:84:  You don't need a ; after a }  [readability/braces] [4]
#10 27.39 src/python/python_api.cpp:367:  Lines should be <= 100 characters long  [whitespace/line_length] [2]
#10 27.39 src/python/python_api.cpp:375:  Lines should be <= 100 characters long  [whitespace/line_length] [2]
#10 27.39 src/python/python_api.cpp:381:  Lines should be <= 100 characters long  [whitespace/line_length] [2]
------
Dockerfile:9
```

**Comment by lloydzhou** [comment]
@knightast 

1. 我其实是希望能帮忙review一下代码，看看这种实现方式是否合理？（我平时写python，所以我自己虽然编译过，也用过，但是不确定是否合理）
2. 如果确定可以这样实现，我后面会解决这个ci报错的问题。

File: deps/build_deps.sh [hunk 1]
```diff
@@ -35,6 +35,7 @@ function BuildPackage {
     mkdir -p ${BUILD_DIR}/${NAME}
     cd ${BUILD_DIR}/${NAME}
     cmake ${DEPS_DIR}/${SOURCE_DIR} \
+        -DCMAKE_CXX_FLAGS="-fPIC -fpic" \
```
Comment (hjk41) [inline deps/build_deps.sh:38]: why are there two -fPIC? typo?

File: include/lgraph/lgraph_galaxy.h [hunk 2]
```diff
@@ -34,7 +34,9 @@ namespace lgraph_api {
  *          performing a lot of writing, but can cause data loss for online processing.
  */
 class Galaxy {
+  protected:
```
Comment (hjk41) [inline include/lgraph/lgraph_galaxy.h:37]: class members are private by default, you don't need to specify

File: src/BuildPythonApi.cmake [hunk 3]
```diff
@@ -1,19 +1,218 @@
+cmake_minimum_required(VERSION 3.1)
+
 find_package(PythonLibs 3 REQUIRED)
 
-set(TARGET_PYTHON_API lgraph_python_api)
+find_path(LEVELDB_INCLUDE_PATH NAMES leveldb/db.h)
```
Comment (hjk41) [inline src/BuildPythonApi.cmake:5]: lgraph.so already have these dependencies, by using target_link_libraries(), cmake will make sure python lib will depend on these libraries, too

File: src/python/README.md [hunk 4]
```diff
@@ -0,0 +1,8 @@
+# 增加Cypher查询
```
Comment (hjk41) [inline src/python/README.md:1]: README.md is supposed to tell people what this module does. You can write these contents in the commit log.

File: src/python/python_api.cpp [hunk 5]
```diff
@@ -37,10 +37,53 @@
...
+  public:
+
+    Galaxy(const std::string& dir, bool durable = false, bool create_if_not_exist = true): lgraph_api::Galaxy(dir, durable, create_if_not_exist) {};
+
+    Galaxy(const std::string& dir, const std::string& user, const std::string& password,
+        bool durable, bool create_if_not_exist): lgraph_api::Galaxy(dir, user, password, durable, create_if_not_exist) {};
+
+    ~Galaxy() { this->Close();};
+
+    std::string Cypher(const std::string& graph, const std::string& script) {
```
Comment (hjk41) [inline src/python/python_api.cpp:69]: Why not implement this function in lgraph_galaxy.cpp?

**Comment by wangtao9** [comment]
> 
> 1. 我其实是希望能帮忙review一下代码，看看这种实现方式是否合理？（我平时写python，所以我自己虽然编译过，也用过，但是不确定是否合理）
> 2. 如果确定可以这样实现，我后面会解决这个ci报错的问题。

很棒的想法，也是很有效的需求。
实现思路需要调整一下，python的lgraph_api应该对应cpp lgraph_api（是其封装），因此如果要提供cypher on lgraph_api功能，应该在cpp lgraph_api中实现，然后封装成python api，而不应该在python_api中实现

**Comment by wangtao9** [comment]
@lloydzhou When implementing "Cypher on C++ lgraph_api", you may need to be careful with nested transactions.

**Comment by qishipengqsp** [comment]
Too old to continue. Recorded in #397. Re-create this pr later @lloydzhou

**Pull request closed without merging**